src/names.c

```diff
@@ -1145,6 +1154,80 @@ void insert_spec(nametab_t *tab, tree_t spec, spec_kind_t kind,
    *p = s;
 }
 
+void set_impl_label_proc_cnt(tree_t t, nametab_t *tab)
+{
+   if (t == NULL) {
+      tab->top_scope->lbl_cnts.proc = 0;
+      return;
+   }
+
+   assert (tree_kind(t) == T_ENTITY);
+
+   const int nstmts = tree_stmts(t);
+   int cnt = 0;
+   ident_t prefix = ident_new("_P");
+   for (int i = 0; i < nstmts; i++) {
+      tree_t stmt = tree_stmt(t, i);
+
+      // Entity statements for sure have labels since they have been implicitly
+      // labelled
+      if (ident_starts_with(tree_ident(stmt), prefix))
+         cnt++;
+   }
+   tab->top_scope->lbl_cnts.proc = cnt;
+}
+
+int* find_impl_label_loop_cnt(nametab_t *tab)
+{
+   scope_t *s = tab->top_scope;
+
+   for (; s != NULL; s = s->parent) {
+      if (s->container == NULL)
+         continue;
+
+      if (is_subprogram(s->container) || tree_kind(s->container) == T_PROCESS)
+         return &(s->lbl_cnts.loop);
+   }
+
+   return NULL;
+}
+
+ident_t get_implicit_label(tree_t t, nametab_t *tab)
+{
+   int *cnt;
+   char c;
+   char buf[22];
+
+   switch (tree_kind(t)) {
+   case T_PROCESS:
+   case T_CONCURRENT:
+      cnt = &(tab->top_scope->lbl_cnts.proc);
+      c = 'P';
+      break;
+
+   case T_FOR:
+   case T_WHILE:
+      cnt = find_impl_label_loop_cnt(tab);
+      c = 'L';
+      break;
+      
+   default:
+      cnt = &(tab->top_scope->lbl_cnts.stmt);
+      c = 'S';
+      break;
+   }
+
+   checked_sprintf(buf, 22, "_%C%x", c, *cnt);
+   (*cnt)++;
+   ident_t ident = ident_new(buf);
+
+#if 0
```

Comment:
Don't leave debugging prints in the code

Reply:
OK, removed. Out of curiosity, what functions should I use for debug logging?

Reply:
You can use printf or whatever, but just take them out when you commit, otherwise it looks messy. Some debug dumping is controlled by the `NVC_*_VERBOSE` environment variables.

Reply:
Thanks, I will keep it in mind.